contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -635,6 +635,33 @@ zfs_key_config_modify_session_counter(pam_handle_t *pamh,
 	return (counter_value);
 }
 
+/* This module only works if every opened session is later closed.  There are
+ * known and common service(s) which violate this norm; warn if they are
+ * detected.
+ */
+static int
+session_roundtrip_check(pam_handle_t *pamh) {
+	const char *service;
+	int ret = pam_get_item(pamh, PAM_SERVICE, (const void **)&service);
+	if (ret == PAM_SUCCESS) {
+		/* See https://github.com/systemd/systemd/issues/8598 */
+		if (strcmp("systemd-user", service) == 0) {
+			/* Workaround:
+			 *	session [success=1 default=ignore] pam_succeed_if.so service = systemd-user
+			 *	session optional                   pam_zfs_key.so
+			 */
+			pam_syslog(pamh, LOG_WARNING,
+				"Service '%s' is known to have problems with sessions",
```

Comment:
I kept this message generic in case other services are found that violate our expectation of 1 session open and 1 session close (perhaps in a different way than `systemd-user`).  However, system administrators might benefit from a more detailed warning.

I'm open to suggestions

Reply:
I think the message should point at either a ZFS issue or a systemd issue where the status can be checked for it being fixed.

Reply:
Some ideas:
- Expect issues; see https://github.com/systemd/systemd/issues/8598
- Session will never close; see https://github.com/systemd/systemd/issues/8598
- Key will not be unloaded because of https://github.com/systemd/systemd/issues/8598

Reply:
Maybe a date? "As of 2022-01-28" + one of your suggested wordings.

I like the "Key will not be unloaded"

Reply:
With a date in the message, we'll likely want to keep the value up-to-date with - at least - yearly commits.  Given the original bug is over 3 years old, and the complexity required for its correct resolution, I expect this would require a long term commitment to update the date.

Let's soften it with "key _may_ not be unloaded", which leaves open the possibility that the issue gets fixed and makes removing the warning message less urgent if a fix for https://github.com/systemd/systemd/issues/8598 does materialize.